File: PlaywireSDKAppsObjC/ads/view/native/CustomNativeAdView.m
//
//  CustomNativeAdView.m
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 01/06/23.
//  Copyright © 2023 Playwire. All rights reserved.
//

#import "CustomNativeAdView.h"

@interface CustomNativeAdView ()
@property (weak, nonatomic) IBOutlet UILabel *headlineLabel;
@property (weak, nonatomic) IBOutlet UILabel *bodyLabel;
@property (weak, nonatomic) IBOutlet UIImageView *iconView;
@property (weak, nonatomic) IBOutlet UILabel *starLabel;
@property (weak, nonatomic) IBOutlet UILabel *priceLabel;
@property (weak, nonatomic) IBOutlet UILabel *storeLabel;
@property (weak, nonatomic) IBOutlet UILabel *advertiserLabel;
@property (weak, nonatomic) IBOutlet UIView *mediaViewHolder;
@property (strong, nonatomic) IBOutlet UIView *contentView;
@end

@implementation CustomNativeAdView

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self setupView];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupView];
    }
    return self;
}

- (instancetype)init {
    self = [self initWithFrame:CGRectZero];
    return self;
}

- (void)setupView {
    [[NSBundle mainBundle] loadNibNamed:@"CustomNativeAdView" owner:self options:nil];
    
    self.contentView.translatesAutoresizingMaskIntoConstraints = false;
    [self addSubview: self.contentView];
    self.bounds = CGRectMake(self.bounds.origin.x, self.bounds.origin.y, self.contentView.bounds.size.width, self.contentView.bounds.size.height);
    [NSLayoutConstraint activateConstraints:@[
        [self.contentView.leftAnchor constraintEqualToAnchor:self.leftAnchor],
        [self.contentView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
        [self.contentView.topAnchor constraintEqualToAnchor:self.topAnchor],
        [self.contentView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor]
    ]];
}

- (void)configure:(PWNativeViewContent *)adContent {
    // Configure views with ad content
    self.iconView.image = adContent.image;
    self.headlineLabel.text = adContent.headline;
    self.actionButton.titleLabel.text = adContent.callToAction;
    self.bodyLabel.text = adContent.body;
    
    if (adContent.starRating) {
        self.starLabel.text = [NSString stringWithFormat:@"%@ ⭐️", [adContent.starRating stringValue]];
    }
    // Hide view in case ad content doesn't contain required information
    [self.starLabel setHidden:adContent.starRating == NULL];

    self.priceLabel.text = adContent.price;
    [self.priceLabel setHidden:adContent.price == NULL];
    
    self.advertiserLabel.text = adContent.advertiser;
    [self.advertiserLabel setHidden:adContent.advertiser == NULL];
    
    self.storeLabel.text = adContent.store;
    [self.storeLabel setHidden:adContent.store == NULL];
    

    UIView* mediaView = adContent.mediaView;
    if (mediaView) {
        mediaView.translatesAutoresizingMaskIntoConstraints = false;
        [self.mediaViewHolder addSubview:mediaView];
        [NSLayoutConstraint activateConstraints:@[
            [mediaView.topAnchor constraintEqualToAnchor: self.mediaViewHolder.topAnchor],
            [mediaView.bottomAnchor constraintEqualToAnchor: self.mediaViewHolder.bottomAnchor],
            [mediaView.leadingAnchor constraintEqualToAnchor: self.mediaViewHolder.leadingAnchor],
            [mediaView.trailingAnchor constraintEqualToAnchor: self.mediaViewHolder.trailingAnchor],
        ]];
    }
}


@end

File: PlaywireSDKAppsSwift/ads/view/native/CustomNativeAdView.swift
//
//  CustomNativeAdView.swift
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 12/20/22.
//  Copyright © 2022 Playwire. All rights reserved.
//

import UIKit
import Playwire

final class CustomNativeAdView: UIView {
    
    @IBOutlet weak var headlineLabel: UILabel!
    @IBOutlet weak var bodyLabel: UILabel!
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var actionButton: UIButton!
    @IBOutlet weak var iconView: UIImageView!
    @IBOutlet weak var starLabel: UILabel!
    @IBOutlet weak var priceLabel: UILabel!
    @IBOutlet weak var storeLabel: UILabel!
    @IBOutlet weak var advertiserLabel: UILabel!
    @IBOutlet weak var mediaViewHolder: UIView!
    @IBOutlet private weak var contentView: UIView!
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupView()
    }
    
    private func setupView() {
        let name = String(describing: CustomNativeAdView.self)
        Bundle.main.loadNibNamed(name, owner: self)
        
        bounds = CGRect(origin: .zero, size: contentView.bounds.size)
        
        contentView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(contentView)
        
        NSLayoutConstraint.activate([
            contentView.leadingAnchor.constraint(equalTo: leadingAnchor),
            contentView.trailingAnchor.constraint(equalTo: trailingAnchor),
            contentView.topAnchor.constraint(equalTo: topAnchor),
            contentView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
    }
    
    func configure(_ adContent: PWNativeViewContent) {
        // Configure views with ad content.
        iconView.image = adContent.icon
        
        headlineLabel.text = adContent.headline
        actionButton.setTitle(adContent.callToAction, for: .normal)
        bodyLabel.text = adContent.body
        
        starLabel.text = adContent.starRating.map { "\($0.stringValue) ⭐️" }
        // Hide view in case ad content doesn't contain required information.
        starLabel.isHidden = adContent.starRating == nil

        priceLabel.text = adContent.price
        priceLabel.isHidden = adContent.price == nil
        
        advertiserLabel.text = adContent.advertiser
        advertiserLabel.isHidden = adContent.advertiser == nil
        
        storeLabel.text = adContent.store
        storeLabel.isHidden = adContent.store == nil
                
        if let mediaView = adContent.mediaView {
            mediaView.translatesAutoresizingMaskIntoConstraints = false
            mediaViewHolder.addSubview(mediaView)
            NSLayoutConstraint.activate([
                mediaView.leadingAnchor.constraint(equalTo: mediaViewHolder.leadingAnchor),
                mediaView.trailingAnchor.constraint(equalTo: mediaViewHolder.trailingAnchor),
                mediaView.topAnchor.constraint(equalTo: mediaViewHolder.topAnchor),
                mediaView.bottomAnchor.constraint(equalTo: mediaViewHolder.bottomAnchor)
            ])
        }
    }
}
